Attack dog leap improvement

forcecore opened the pull request:
Attempt to fix #7506

Based on Mailaender's dog improvement PR #11715. (Sorry about C&P instead of Git Cherry Pick but it was way behind bleed)

Demo:

* https://www.youtube.com/watch?v=GLGbeL4F-fg : unhardcoded 1-shot kill
* https://www.youtube.com/watch?v=BS7e8b99L-w : Dog fight (no more indefinite jump loop)
* https://www.youtube.com/watch?v=eKaIIwWaye0 : vs Riflemen

forcecore committed:
Attack dog leap improvement

pchote commented on OpenRA.Mods.Cnc/Activities/LeapAttack.cs:
```diff
@@ -0,0 +1,112 @@
+
+			var leapToken = ConditionManager.InvalidConditionToken;
+
+			return ActivityUtils.SequenceActivities(
```
Please use child activities and OnFirstRun instead of SequenceActivities.  Refer to the Transform activity for an example.

forcecore commented:
* Now using parent-child activity
* Buff on and off now encapsulated in AttackLeap class.
Leap stuff shouldn't be pre-calculated in OnFirstRun otherwise it will be incorrect, as self can change position.

forcecore commented:
Forgot to mention, in parent-child activity version, I addressed far-jumping dogs
https://youtu.be/1lGmjOaNBxQ

forcecore commented:
Fixed shell map crash that wasn't caught before

reaperrr commented:
Needs rebase.

forcecore commented:
Rebased

atlimit8 commented on OpenRA.Mods.Cnc/Activities/Leap.cs:
```diff
@@ -9,70 +9,57 @@
+		}
 	}
-}
+}
```
nit: Leave newline here.

atlimit8 commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,71 @@
+				conditionManager.GrantCondition(self, info.AttackingCondition, info.AttackingDuration);
+		}
+	}
+}
```
same here

forcecore commented:
EOL fixed, squashed.

atlimit8 commented on OpenRA.Mods.Cnc/Activities/LeapAttack.cs:
```diff
@@ -0,0 +1,118 @@
+			{
+				attack.ApproachBuffOn(self);
+				QueueChild(new MoveWithinRange(self, target, WDist.Zero, range));
+				QueueChild(new CallFunc(() => attack.ApproachBuffOff(self)));
```
Remove:
```c#
				QueueChild(new CallFunc(() => attack.ApproachBuffOff(self)));
```

atlimit8 commented on OpenRA.Mods.Cnc/Activities/LeapAttack.cs:
```diff
@@ -0,0 +1,118 @@
+			var range = attack.GetMaximumRange();
+			if (!target.IsInRange(self.CenterPosition, range))
+			{
+				attack.ApproachBuffOn(self);
```
Append:
```c#
				isApproachBuffOn = true;
```

atlimit8 commented on OpenRA.Mods.Cnc/Activities/LeapAttack.cs:
```diff
@@ -0,0 +1,118 @@
+
+			if (ChildActivity != null)
+			{
+				ActivityUtils.RunActivity(self, ChildActivity);
```
Replace with:
```c#
				var act = ActivityUtils.RunActivity(self, ChildActivity);
				if (isApproachBuffOn && act == this))
					attack.ApproachBuffOff(self);
```

atlimit8 commented on OpenRA.Mods.Cnc/Activities/LeapAttack.cs:
```diff
@@ -0,0 +1,118 @@
+		readonly Target target;
+		readonly AttackLeapInfo info;
+		readonly Mobile mobile;
+		readonly AttackLeap attack;
```
Append:
```c#
		bool isApproachBuffOn = false;
```

atlimit8 reviewed (commented):
The`CallFunc` activity is not serialization friendly (game saving). Granted, it may take a while before that can be added; however, it is better to code for its support now than to make more work for later.

forcecore committed:
(to be squashed) YAML update for bleed, visual glitch removal, CallFunc removal

forcecore commented:
Removed CallFunc. Removed more visual glitch.

	AttackLeap:
		Voice: Attack
		LeapCondition: leap
		ApproachCondition: rush
		# Disable another leap while leaping or ripping
		RequiresCondition: !leap && !rip

I also made rip condition to disable AttackLeap to prevent dogs from jumping immediately. The problem is, if I queue attack orders, rip disables the attack trait and the queue gets canceled. I didn't look into detail on the reason but I can't queue with && !rip condition.

forcecore committed:
More simplification

pchote commented:
> I think AttackBase trait should be a pausable trait as well.

This sounds like a sensible solution.  Would you mind filing a PR for this?

pchote commented:
This reintroduces problems with dogs stacking on top of other infantry.

Repro case:
* Start a game with two clients
* Player one builds a cell full of non-combat infantry, or normal infantry in the hold fire stance
* Player two builds 5 dogs, puts them in the hold fire stance, and orders them to move next to player 1's units.
* Player two selects all 5 dogs and orders them to attack the center unit from Player one's cell.

The dogs will then leap but not adjust their sub-cells. The outer 4 will be left overlapping with Player one's other infantry:
![screen shot 2017-08-17 at 18 03 57](https://user-images.githubusercontent.com/167819/29424265-eaf898b4-8376-11e7-8dbf-63f153320039.png)

GraionDilach commented:
I'd like to see this trait renamed to AttackJoust during this PR, because it no longer involves a leap.

See my reasoning at https://github.com/OpenRA/ra2/issues/328#issuecomment-277515557.

forcecore commented:
I let them overlap. Otherwise, when the cell contains 5 enemy infantries, attack dogs can't attack any of those infantries at all. If I don't hard-code one-shot kill into the code, I'd get overlap anyway.

forcecore commented:
>> I think AttackBase trait should be a pausable trait as well.
> This sounds like a sensible solution. Would you mind filing a PR for this?

I'll try. That should improve EMP in TS too.

pchote commented:
> If I don't hard-code one-shot kill into the code, I'd get overlap anyway if target HP > dog damage. 

This is precisely the reason for the one-shot kill.

GraionDilach commented:
Hardcoding the one-shot-kill will remove the chance of using this implementation for parasite behaviour (RA2 Terror Drone vs vehicles, Firestorm Limpet Drone). Another option would be that the actor would bounce to a free subcell after an attack when the target isn't killed.

forcecore commented:
Hmm, probably why the attack dogs in RA1 "slides" after ripping up. I'll try that.

pchote commented:
> I think the proper way to avoid this is to introduce some kind of a coordination to the jumping, something like resource claim layer. What do you think about the doggy jumping coordination layer?

Yes, we need this, but not as a layer.  Implement it as a trait on the actor itself, like we have for external capturing or the various things that lock buildings.

pchote commented:
> Hardcoding the one-shot-kill will remove the chance of using this implementation for parasite behaviour

Parasite behaviour requires the targeting actor to enter the target, which is a completely different usecase.  Trying to solve that with the same code as attack dogs makes for uglier and more complicated solutions than necessary for *both*.

GraionDilach commented:
Squid parasite didn't really entered into the target, was moreso attached to.

pchote commented:
Squid parasite behaviour requires the targeting actor to attach to the target, which is a completely different usecase. Trying to solve that with the same code as attack dogs makes for uglier and more complicated solutions than necessary for both.

:wink:

pchote commented:
@forcecore: do you plan on continuing with this in the short term?

forcecore commented:
Not in the short term, I have personal issues to sort out for about a few weeks.

pchote commented:
Ok, no worries.  I've moved this to the Future milestone so that we don't lose track of this, and will close it for now to help reduce the PR queue to items that need active review.
